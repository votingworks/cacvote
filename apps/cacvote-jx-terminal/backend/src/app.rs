//! Application definition, including all HTTP route handlers.
//!
//! Route handlers are bundled via [`setup`] into an [`axum::Router`], which can then be run
//! using [`run`] at the configured port (see [`config`][`super::config`]).

use std::convert::Infallible;
use std::net::{IpAddr, Ipv4Addr, SocketAddr};
use std::time::Duration;

use async_stream::try_stream;
use auth_rs::card_details::CardDetailsWithAuthInfo;
use axum::body::Bytes;
use axum::response::sse::{Event, KeepAlive};
use axum::response::Sse;
use axum::routing::post;
use axum::Json;
use axum::{extract::DefaultBodyLimit, routing::get, Router};
use axum::{extract::State, http::StatusCode, response::IntoResponse};
use futures_core::Stream;
use serde_json::json;
use sqlx::PgPool;
use tokio::time::sleep;
use tower_http::services::{ServeDir, ServeFile};
use tower_http::trace::TraceLayer;
use tracing::Level;
use types_rs::cacvote::{AuthStatus, Election, Payload, SessionData, SignedObject};
use types_rs::election::ElectionDefinition;
use uuid::Uuid;

use crate::config::{Config, MAX_REQUEST_SIZE};
use crate::{db, smartcard};

// type AppState = (Config, PgPool, smartcard::StatusGetter);
type AppState = (Config, PgPool, smartcard::DynSmartcard);

/// Prepares the application with all the routes. Run the application with
/// `app::run(…)` once you have it.
pub(crate) fn setup(pool: PgPool, config: Config, smartcard: smartcard::DynSmartcard) -> Router {
    let _entered = tracing::span!(Level::DEBUG, "Setting up application").entered();

    let router = match &config.public_dir {
        Some(public_dir) => Router::new().fallback_service(
            ServeDir::new(public_dir)
                .append_index_html_on_directories(true)
                .fallback(ServeFile::new(public_dir.join("index.html"))),
        ),
        None => {
            tracing::info!("No PUBLIC_DIR configured, serving no files");
            Router::new()
        }
    };

    router
        .route("/api/status", get(get_status))
        .route("/api/status-stream", get(get_status_stream))
        .route("/api/elections", get(get_elections))
        .route("/api/elections", post(create_election))
        .layer(DefaultBodyLimit::max(MAX_REQUEST_SIZE))
        .layer(TraceLayer::new_for_http())
        .with_state((config, pool, smartcard))
}

/// Runs an application built by `app::setup(…)`.
pub(crate) async fn run(app: Router, config: &Config) -> color_eyre::Result<()> {
    let addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), config.port);
    tracing::info!("Server listening at http://{addr}/");
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await?;
    Ok(())
}

async fn get_status() -> impl IntoResponse {
    StatusCode::OK
}

async fn get_status_stream(
    State((config, _, smartcard)): State<AppState>,
) -> Sse<impl Stream<Item = Result<Event, Infallible>>> {
    Sse::new(try_stream! {
        let mut last_card_details = None;

        loop {
            let new_card_details = smartcard.get_card_details();

            if new_card_details != last_card_details {
                last_card_details = new_card_details.clone();
                yield Event::default().json_data(SessionData {
                    auth_status: match new_card_details {
                        Some(CardDetailsWithAuthInfo { card_details, .. }) if card_details.jurisdiction_code() == config.jurisdiction_code => AuthStatus::Authenticated,
                        Some(_) => AuthStatus::UnauthenticatedInvalidCard,
                        None => AuthStatus::UnauthenticatedNoCard,
                    },
                    jurisdiction_code: Some(config.jurisdiction_code.clone()),
                }).unwrap();
            }

            sleep(Duration::from_millis(100)).await;
        }
    })
    .keep_alive(KeepAlive::default())
}

async fn get_elections(State((_, pool, _)): State<AppState>) -> impl IntoResponse {
    let mut connection = match pool.acquire().await {
        Ok(connection) => connection,
        Err(e) => {
            tracing::error!("error getting database connection: {e}");
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "error getting database connection" })),
            );
        }
    };

    let elections = match db::get_elections(&mut connection).await {
        Ok(elections) => elections,
        Err(e) => {
            tracing::error!("error getting elections from database: {e}");
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "error getting elections from database" })),
            );
        }
    };

    (StatusCode::OK, Json(json!({ "elections": elections })))
}

async fn create_election(
    State((_, pool, smartcard)): State<AppState>,
    body: Bytes,
) -> impl IntoResponse {
    let jurisdiction_code = match smartcard.get_card_details() {
        Some(card_details) => card_details.card_details.jurisdiction_code(),
        None => {
            tracing::error!("no card details found");
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "no card details found" })),
            );
        }
    };

    let election_definition = match ElectionDefinition::try_from(&body[..]) {
        Ok(election_definition) => election_definition,
        Err(e) => {
            tracing::error!("error parsing election definition: {e}");
            return (
                StatusCode::BAD_REQUEST,
                Json(json!({ "error": format!("error parsing election definition: {e}") })),
            );
        }
    };

    let mut connection = match pool.acquire().await {
        Ok(connection) => connection,
        Err(e) => {
            tracing::error!("error getting database connection: {e}");
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "error getting database connection" })),
            );
        }
    };

    let payload = Payload::Election(Election {
        jurisdiction_code,
        election_definition,
    });
    let serialized_payload = match serde_json::to_vec(&payload) {
        Ok(serialized_payload) => serialized_payload,
        Err(e) => {
            tracing::error!("error serializing payload: {e}");
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "error serializing payload" })),
            );
        }
    };

    let signed = match smartcard.sign(&serialized_payload, None) {
        Ok(signed) => signed,
        Err(e) => {
            tracing::error!("error signing payload: {e}");
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "error signing payload" })),
            );
        }
    };
    let certificates: Vec<u8> = match signed
        .cert_stack
        .iter()
        .map(|cert| cert.to_pem())
        .collect::<Result<Vec<_>, _>>()
    {
        Ok(certificates) => certificates.concat(),
        Err(e) => {
            tracing::error!("error converting certificates to PEM: {e}");
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "error converting certificates to PEM" })),
            );
        }
    };
    let signed_object = SignedObject {
        id: Uuid::new_v4(),
        payload: serialized_payload,
        certificates,
        signature: signed.data,
    };

    if let Err(e) = db::add_object(&mut connection, &signed_object).await {
        tracing::error!("error adding object to database: {e}");
        return (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({ "error": "error adding object to database" })),
        );
    }

    (StatusCode::CREATED, Json(json!({ "id": signed_object.id })))
}
