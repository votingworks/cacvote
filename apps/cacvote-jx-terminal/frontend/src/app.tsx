import { QueryClientProvider } from '@tanstack/react-query';
import { AppBase, ErrorBoundary, P } from '@votingworks/ui';
import { BrowserRouter } from 'react-router-dom';
import { createQueryClient } from './api';
import { AppRoot } from './app_root';

export function App(): JSX.Element {
  const queryClient = createQueryClient();

  return (
    <AppBase
      defaultColorMode="desktop"
      defaultSizeMode="desktop"
      screenType="builtIn"
    >
      <ErrorBoundary errorMessage={<P>Something went wrong</P>}>
        <QueryClientProvider client={queryClient}>
          <BrowserRouter>
            <AppRoot />
          </BrowserRouter>
        </QueryClientProvider>
      </ErrorBoundary>
    </AppBase>
  );
}
