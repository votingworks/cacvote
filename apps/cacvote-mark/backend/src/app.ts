import { cac } from '@votingworks/auth';
import {
  err,
  ok,
  Optional,
  Result,
  throwIllegalValue,
} from '@votingworks/basics';
import * as grout from '@votingworks/grout';
import {
  BallotStyleId,
  ElectionDefinition,
  PrecinctId,
  unsafeParse,
  VotesDict,
} from '@votingworks/types';
import express, { Application } from 'express';
import { isDeepStrictEqual } from 'util';
import { v4 } from 'uuid';
import { DateTime } from 'luxon';
import { Auth, AuthStatus } from './types/auth';
import { Workspace } from './workspace';
import {
  JurisdictionCode,
  Payload,
  RegistrationRequest,
  RegistrationRequestObjectType,
  SignedObject,
  Uuid,
  UuidSchema,
} from './cacvote-server/types';

export type VoterStatus =
  | 'unregistered'
  | 'registration_pending'
  | 'registered'
  | 'voted';

function buildApi({
  auth,
  workspace: { store },
}: {
  auth: Auth;
  workspace: Workspace;
}) {
  async function getAuthStatus(): Promise<AuthStatus> {
    return await auth.getAuthStatus();
  }

  return grout.createApi({
    getAuthStatus,

    checkPin(input: { pin: string }) {
      return auth.checkPin(input.pin);
    },

    getJurisdictionsCodes() {
      return store.getJurisdictionCodes();
    },

    async getVoterStatus(): Promise<Optional<{ status: VoterStatus }>> {
      const authStatus: AuthStatus = await getAuthStatus();

      if (authStatus.status !== 'has_card') {
        return undefined;
      }

      const { commonAccessCardId } = authStatus.card;

      // TODO: support more than one registration request for a given voter
      const registrationRequest = store
        .forEachRegistrationRequest({ commonAccessCardId })
        .first();

      if (!registrationRequest) {
        return { status: 'unregistered' };
      }

      const registration = store
        .forEachRegistration({
          commonAccessCardId,
          registrationRequestObjectId: registrationRequest.object.getId(),
        })
        .first();

      if (!registration) {
        return { status: 'registration_pending' };
      }

      // TODO: check for a submitted ballot to see if the voter has voted
      return { status: 'registered' };
    },

    async getRegistrationRequests(): Promise<RegistrationRequest[]> {
      const authStatus = await getAuthStatus();

      if (authStatus.status !== 'has_card') {
        return [];
      }

      // TODO: get registration requests for the user
      return [];
    },

    async createVoterRegistrationRequest(input: {
      jurisdictionCode: JurisdictionCode;
      givenName: string;
      familyName: string;
      pin: string;
    }): Promise<
      Result<
        { id: Uuid },
        { type: 'not_logged_in' | 'incorrect_pin'; message: string }
      >
    > {
      const authStatus = await getAuthStatus();

      if (authStatus.status !== 'has_card') {
        return err({ type: 'not_logged_in', message: 'Not logged in' });
      }

      if (!(await auth.checkPin(input.pin))) {
        return err({ type: 'incorrect_pin', message: 'Incorrect PIN' });
      }

      const registrationRequest = new RegistrationRequest(
        authStatus.card.commonAccessCardId,
        input.jurisdictionCode,
        input.givenName,
        input.familyName,
        DateTime.now()
      );

      const payload = Payload.of(
        RegistrationRequestObjectType,
        registrationRequest
      ).toBuffer();
      const generateSignatureResult = await auth.generateSignature(payload, {
        pin: input.pin,
      });

      if (generateSignatureResult.isErr()) {
        const error = generateSignatureResult.err();
        switch (error.type) {
          case 'card_error':
            return err({
              type: 'not_logged_in',
              message: `Card error: ${generateSignatureResult.err().message}`,
            });

          case 'incorrect_pin':
            return err({ type: 'incorrect_pin', message: 'Incorrect PIN' });

          default:
            throwIllegalValue(error);
        }
      }

      const certificates = await auth.getCertificate();
      const objectId = unsafeParse(UuidSchema, v4());
      const object = new SignedObject(
        objectId,
        payload,
        certificates,
        generateSignatureResult.ok()
      );

      (await store.addObject(object)).unsafeUnwrap();

      return ok({ id: objectId });
    },

    async getElectionConfiguration(): Promise<
      Optional<{
        electionDefinition: ElectionDefinition;
        ballotStyleId: BallotStyleId;
        precinctId: PrecinctId;
      }>
    > {
      const authStatus = await getAuthStatus();

      if (authStatus.status !== 'has_card') {
        return undefined;
      }

      // TODO: get election configuration for the user
      return undefined;
    },

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    castBallot(_input: {
      votes: VotesDict;
      pin: string;
    }): Promise<Result<Uuid, cac.GenerateSignatureError>> {
      // TODO: cast and print the ballot
      throw new Error('Not implemented');
    },

    logOut() {
      return auth.logOut();
    },
  });
}

export type Api = ReturnType<typeof buildApi>;

export function buildApp({
  auth,
  workspace,
}: {
  auth: Auth;
  workspace: Workspace;
}): Application {
  const app: Application = express();
  const api = buildApi({ auth, workspace });

  app.use('/api/watchAuthStatus', (req, res) => {
    res.set({
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      Connection: 'keep-alive',
    });

    let timeout: NodeJS.Timeout | undefined;
    let lastAuthStatus: AuthStatus | undefined;

    async function sendUpdate() {
      const authStatus = await api.getAuthStatus();

      if (!isDeepStrictEqual(authStatus, lastAuthStatus)) {
        lastAuthStatus = authStatus;
        res.write(`data: ${grout.serialize(authStatus)}\n\n`);
      }

      timeout = setTimeout(
        sendUpdate,
        10 /* AUTH_STATUS_POLLING_INTERVAL_MS */
      );
    }

    req.on('close', () => {
      clearTimeout(timeout);
      res.end();
    });

    void sendUpdate();
  });

  app.use('/api', grout.buildRouter(api, express));
  return app;
}
