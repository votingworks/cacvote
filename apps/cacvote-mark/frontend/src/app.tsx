import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { LogSource, Logger } from '@votingworks/logging';
import { ColorMode, ScreenType, SizeMode } from '@votingworks/types';
import { AppBase, ErrorBoundary, H1, P, Prose } from '@votingworks/ui';
import {
  ApiClient,
  ApiClientContext,
  createApiClient,
  createQueryClient,
} from './api';
import { AppRoot } from './app_root';
import { DisplaySettingsManager } from './components/display_settings_manager';

const DEFAULT_COLOR_MODE: ColorMode = 'contrastMedium';
const DEFAULT_SCREEN_TYPE: ScreenType = 'elo15';
const DEFAULT_SIZE_MODE: SizeMode = 'touchSmall';

export interface Props {
  logger?: Logger;
  apiClient?: ApiClient;
  queryClient?: QueryClient;
}

export function App({
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  logger = new Logger(LogSource.VxMarkFrontend),
  /* istanbul ignore next */ apiClient = createApiClient(),
  queryClient = createQueryClient(),
}: Props): JSX.Element {
  return (
    <AppBase
      defaultColorMode={DEFAULT_COLOR_MODE}
      defaultSizeMode={DEFAULT_SIZE_MODE}
      screenType={DEFAULT_SCREEN_TYPE}
    >
      <ErrorBoundary
        errorMessage={
          <Prose textCenter>
            <H1>Something went wrong</H1>
            <P>Ask a poll worker to restart the ballot marking device.</P>
          </Prose>
        }
      >
        <ApiClientContext.Provider value={apiClient}>
          <QueryClientProvider client={queryClient}>
            <DisplaySettingsManager />
            <AppRoot />
          </QueryClientProvider>
        </ApiClientContext.Provider>
      </ErrorBoundary>
    </AppBase>
  );
}
