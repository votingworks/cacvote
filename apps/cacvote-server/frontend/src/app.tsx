import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ColorMode, ScreenType, SizeMode } from '@votingworks/types';
import { AppBase, ErrorBoundary, H1, P, Prose } from '@votingworks/ui';
import { BrowserRouter } from 'react-router-dom';
import { createQueryClient } from './api';

import { AppRoot } from './app_root';

const DEFAULT_COLOR_MODE: ColorMode = 'contrastMedium';
const DEFAULT_SCREEN_TYPE: ScreenType = 'elo15';
const DEFAULT_SIZE_MODE: SizeMode = 'touchSmall';

export interface Props {
  queryClient?: QueryClient;
}

export function App({ queryClient = createQueryClient() }: Props): JSX.Element {
  return (
    <AppBase
      defaultColorMode={DEFAULT_COLOR_MODE}
      defaultSizeMode={DEFAULT_SIZE_MODE}
      screenType={DEFAULT_SCREEN_TYPE}
    >
      <ErrorBoundary
        errorMessage={
          <Prose textCenter>
            <H1>Something went wrong</H1>
            <P>Please reload the page.</P>
          </Prose>
        }
      >
        <QueryClientProvider client={queryClient}>
          <BrowserRouter>
            <AppRoot />
          </BrowserRouter>
        </QueryClientProvider>
      </ErrorBoundary>
    </AppBase>
  );
}
