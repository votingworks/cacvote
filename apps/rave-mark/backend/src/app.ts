import { useDevDockRouter } from '@votingworks/dev-dock-backend';
import * as grout from '@votingworks/grout';
import express, { Application } from 'express';

function buildApi() {
  return grout.createApi({});
}

export type Api = ReturnType<typeof buildApi>;

export function buildApp(): Application {
  const app: Application = express();
  const api = buildApi();
  app.use('/api', grout.buildRouter(api, express));
  useDevDockRouter(app, express);
  return app;
}
