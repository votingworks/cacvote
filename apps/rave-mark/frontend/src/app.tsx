import React from 'react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { LogSource, Logger } from '@votingworks/logging';
import { ColorMode, ScreenType, SizeMode } from '@votingworks/types';
import { AppBase, ErrorBoundary, H1, P, Prose } from '@votingworks/ui';
import { BrowserRouter } from 'react-router-dom';
import {
  ApiClient,
  ApiClientContext,
  createApiClient,
  createQueryClient,
} from './api';

const DEFAULT_COLOR_MODE: ColorMode = 'contrastMedium';
const DEFAULT_SCREEN_TYPE: ScreenType = 'elo15';
const DEFAULT_SIZE_MODE: SizeMode = 'm';

export interface Props {
  logger?: Logger;
  apiClient?: ApiClient;
  queryClient?: QueryClient;
}

export function App({
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  logger = new Logger(LogSource.VxMarkFrontend),
  /* istanbul ignore next */ apiClient = createApiClient(),
  queryClient = createQueryClient(),
}: Props): JSX.Element {
  return (
    <AppBase
      defaultColorMode={DEFAULT_COLOR_MODE}
      defaultSizeMode={DEFAULT_SIZE_MODE}
      isTouchscreen
      screenType={DEFAULT_SCREEN_TYPE}
    >
      <BrowserRouter>
        <ErrorBoundary
          errorMessage={
            <Prose textCenter>
              <H1>Something went wrong</H1>
              <P>Ask a poll worker to restart the ballot marking device.</P>
            </Prose>
          }
        >
          <ApiClientContext.Provider value={apiClient}>
            <QueryClientProvider client={queryClient}>
              <H1>Hello RAVE</H1>
            </QueryClientProvider>
          </ApiClientContext.Provider>
        </ErrorBoundary>
      </BrowserRouter>
    </AppBase>
  );
}
