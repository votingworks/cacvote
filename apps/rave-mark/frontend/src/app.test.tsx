import React from 'react';
import { render } from '@testing-library/react';
import { App } from './app';

test('renders', () => {
  const { container } = render(<App />);
  expect(container).toMatchSnapshot();
});
