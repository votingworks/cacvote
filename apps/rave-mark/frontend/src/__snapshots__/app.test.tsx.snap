// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <h1
    class="sc-eCImvq dRIUix"
  >
    Hello RAVE
  </h1>
</div>
`;
